### PR TITLE
Merge 3.0 to 3.1

### PR DESCRIPTION
**Pull request by hmlanigan**
Contains:

* #15099 from hmlanigan/lp1999060-single-charm-multi-app
* #15102 from SimonRichardson/pass-series-in-upgrade-machine-charm-hook
* #15111 from jack-w-shaw/cherry-pick-ck-fix-3.0

Merge conflicts with:
* apiserver/facades/controller/charmdownloader/charmdownloader_test.go 
* state/export_test.go 
* state/migration_import.go 
* state/migration_import_test.go
* tests/suites/ck/ck.sh

### CODE REVIEW

**Review by SimonRichardson — CHANGES_REQUESTED** [review]
Something is wrong with the merge. (see the linters)

File: state/upgrades.go [hunk 1]
```diff
@@ -95,3 +94,62 @@ func applyToAllModelSettings(st *State, change func(*settingsDoc) (bool, error))
 	}
 	return nil
 }
+
+// CorrectCharmOriginsMultiAppSingleCharm corrects application charm origins
+// where the charm has been downloaded, however the origin was not updated
+// with ID and Hash data. Per LP 1999060, usually seen with multi application
+// using same charm bundles.
+func CorrectCharmOriginsMultiAppSingleCharm(pool *StatePool) error {
```
Comment (wallyworld) [inline state/upgrades.go:102]: As mentioned in the PR text, this needs to be removed.

**Comment by jack-w-shaw** [comment]
I've just merged https://github.com/juju/juju/pull/15111 into 3.0. While you're at it, it would be great if that could be included in this merge

**Comment by hmlanigan** [comment]
Added @jack-w-shaw pr to merge, reverted the upgrade steps, redo on a charmdownloader test merge for no macaroons.  Also rebased to the current 3.1

Not sure why I'm getting linter issues on files not changed. Will look after the lint run on current code.

**Review by SimonRichardson — APPROVED** [review]
OK to land if the linters say so.

**Review by jack-w-shaw — APPROVED** [review]

**Comment by hmlanigan** [comment]
When I try to regenerate `core/charm/downloader/mocks/downloader_mocks.go`, no changes are shown. `check-multi-juju` is successful too, the tests run.

**Pull request merged**